File: 543.diameter-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func diameterOfBinaryTree(root *TreeNode) int {
	if root == nil {
		return 0
	}

	res := 0
	depth(root, &res)

	return res
}

func depth(root *TreeNode, res *int) int {
	if root == nil {
		return 0
	}

	left := depth(root.Left, res)
	right := depth(root.Right, res)

	*res = max(*res, left + right)

	return max(left, right) + 1
}

func max(a int, b int) int {
	if a > b {
		return a
	}

	return b
}

File: 543.diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int res;

    int diameterOfBinaryTree(TreeNode* root) {
        res = 0;
        diameterOfBinaryTreeHelp(root);
        return res;
    }

    int diameterOfBinaryTreeHelp(TreeNode* root) {
        if (root == NULL) {
            return 0;
        }

        int l = diameterOfBinaryTreeHelp(root->left);
        int r = diameterOfBinaryTreeHelp(root->right);

        res = max(res, l+r);
        return max(l, r) + 1;
    }
};
